backend/infrahub/core/validators/checks_runner.py

```diff
@@ -0,0 +1,35 @@
+import asyncio
+from typing import Coroutine
+
+from infrahub_sdk.node import InfrahubNode
+
+from infrahub.core.constants import ValidatorConclusion, ValidatorState
+from infrahub.core.timestamp import Timestamp
+
+
+async def run_checks_and_update_validator(checks: list[Coroutine], validator: InfrahubNode) -> None:
+    """
+    Execute a list of checks coroutines, and set validator fields accordingly.
+    Tasks are retrieved by completion order so as soon as we detect a failing check,
+    we set validator conclusion to failure.
+    """
+
+    # First set validator to in progress, then wait for results
+    validator.state.value = ValidatorState.IN_PROGRESS.value
+    validator.started_at.value = Timestamp().to_string()
+    validator.completed_at.value = ""
+    await validator.save()
+
+    for earliest_task in asyncio.as_completed(checks):
+        result = await earliest_task
+        if result == ValidatorConclusion.FAILURE:
+            validator.conclusion.value = ValidatorConclusion.FAILURE.value
+            await validator.save()
```

Comment:
If there are lots of artifacts that are failing this extra save could potentially generate a lot of extra noise.At the same time it's nice with an early failure if there are failures. Perhaps we can switch a boolean value on the first failure, then for other following failures we don't try to overwrite the same status if a failure has already occurred.

Reply:
Right!